docs/proposals/localDisk_volume.md

```diff
@@ -0,0 +1,309 @@
+      - "3600"
+    imagePullPolicy: IfNotPresent
+    name: ld-test
+    volumeMounts:
+    - mountPath: /mnt/disk1
+      name: test-vol1
+    - mountPath: /mnt/disk2
+      name: test-vol2
+  volumes:
+  - localDisk:
+        type: disk
+    name: test-vol1
+  - localDisk:
+        type: disk
+    name: test-vol2
+  restartPolicy: Always
+```
+
+2. The pod need a N GigaBytes lvm volume, e.g. 500 GB
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: ld-test
+spec:
+  containers:
+  - image: busybox
+    command:
+      - sleep
+      - "3600"
+    imagePullPolicy: IfNotPresent
+    name: ld-test
+    volumeMounts:
+    - mountPath: /mnt/vol1
+      name: test-vol
+  volumes:
+  - localDisk:
+        volumesize: 500000
+        type: lvm
+    name: test-vol
+  restartPolicy: Always
+```
+
+3. Create a LVC, then create a pod by using this LVC
+
+```yaml
+kind: LocalVolumeClaim
+apiVersion: extensions/v1beta1
+metadata:
+  name: myclaim
+spec:
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 100Gi
+  volumetype: disk    
+```
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: ld-test
+spec:
+  containers:
+  - image: busybox
+    command:
+      - sleep
+      - "3600"
+    imagePullPolicy: IfNotPresent
+    name: ld-test
+    volumeMounts:
+    - mountPath: /mnt/vol
+      name: test-vol
+  volumes:
+  - localVolumeClaim:
+        claimName: myclaim
+    name: test-vol
+  restartPolicy: Always
+```
+
+## Proposed Design
+
+### localDisk volume
+
+New volume type localDisk will be created and registered as volume plugin in kubelet, `pkg/volume/local_disk/local_disk.go`
+
+```go
+type localDiskPlugin struct {
+	host volume.VolumeHost
+	// decouple creating Recyclers/Deleters/Provisioners by deferring to a function.  Allows for easier testing.
+	newRecyclerFunc    func(pvName string, spec *volume.Spec, host volume.VolumeHost, volumeConfig volume.VolumeConfig) (volume.Recycler, error)
+	newDeleterFunc     func(spec *volume.Spec, host volume.VolumeHost) (volume.Deleter, error)
+	newProvisionerFunc func(options volume.VolumeOptions, host volume.VolumeHost) (volume.Provisioner, error)
+	config             volume.VolumeConfig
+}
+
+const (
+	localDiskPluginName = "kubernetes.io/local-disk"
+)
+
+```
+
+In PodSpec, we shall add a new type of VolumeSource, `LocalDiskVolumeSource`
+
+```go
+type LocalDiskVolumeSource struct {
+	// Node and Path here are for localDisk volume recycle pod only
+	Node string `json: node, omitempty`
+	Path string `json:"path, omitempty"`
+	//The size of local disk, unit is MB
+	VolumeSize int64 `json:"volumeszie, omitempty"`
+	// The type of local disk, it could be "disk", "lvm" now.
+	Type LocalDiskType `json:"type"`
+	// The filesystem type of local disk, the default is ext4
+	FSType string `json:"fsType,omitempty"`
+}
+```
+
+### LocalVolume and LocalVolumeClaim (LV/LVC)
+
+The LocalVolume and LocalVolumeClaim (LV/LVC) are new volume management objects in this proposal.

```

Comment:
Given the general overhead this imposes to users (people have to write app templates that work for both types), I think this is the wrong abstraction.  I believe that local volumes are an implementation detail of all persistent volumes, rather than a new top level concept.

1. PersistentVolumes already have the concept of being "location aware" (AWS/GCE volumes are region only)
2. PersistentVolumes already support hostPath which by definition are local

A local volume is a combination of 1 and 2 above.  

Furthermore, creating an entirely different code structure for LV over PV is an enormous cost for internal code structure, security, compatibility, and consistency.  Features added to PV have to be added to LV (quota, limits), and exposed to end users and admins in a way that makes their relationship consistent.

We need a very strong benefit articulated to having LV/LVC separate for PV/PVC.  The reasoning here I don't believe is sufficient yet to hit that bar.